src/connector/src/sink/big_query.rs

```diff
@@ -0,0 +1,384 @@
+        let mut rs = client
+        .job()
+        .query(
+            &self.config.common.project,
+            QueryRequest::new(format!(
+                "SELECT column_name, data_type FROM `{}.{}.INFORMATION_SCHEMA.COLUMNS` WHERE table_name = '{}'"
+                ,self.config.common.project,self.config.common.dataset,self.config.common.table,
+            )),
+        )
+        .await.map_err(|e| SinkError::BigQuery(e.to_string()))?;
+        let mut big_query_schema = HashMap::default();
+        while rs.next_row() {
+            big_query_schema.insert(
+                rs.get_string_by_name("column_name")
+                    .map_err(|e| SinkError::BigQuery(e.to_string()))?
+                    .ok_or_else(|| SinkError::BigQuery("Cannot find column_name".to_owned()))?,
+                rs.get_string_by_name("data_type")
+                    .map_err(|e| SinkError::BigQuery(e.to_string()))?
+                    .ok_or_else(|| SinkError::BigQuery("Cannot find column_name".to_owned()))?,
+            );
+        }
+
+        self.check_column_name_and_type(big_query_schema)?;
+        Ok(())
+    }
+
+    fn default_sink_decouple(_desc: &super::catalog::desc::SinkDesc) -> bool {
+        false
+    }
+
+    async fn new_coordinator(&self) -> Result<Self::Coordinator> {
+        Err(SinkError::Coordinator(anyhow!("no coordinator")))
+    }
+}
+
+pub struct BigQuerySinkWriter {
+    pub config: BigQueryConfig,
+    schema: Schema,
+    pk_indices: Vec<usize>,
+    client: Client,
+    is_append_only: bool,
+    insert_request: TableDataInsertAllRequest,
+    row_encoder: JsonEncoder,
+}
+
+impl TryFrom<SinkParam> for BigQuerySink {
+    type Error = SinkError;
+
+    fn try_from(param: SinkParam) -> std::result::Result<Self, Self::Error> {
+        let schema = param.schema();
+        let config = BigQueryConfig::from_hashmap(param.properties)?;
+        BigQuerySink::new(
+            config,
+            schema,
+            param.downstream_pk,
+            param.sink_type.is_append_only(),
+        )
+    }
+}
+
+impl BigQuerySinkWriter {
+    pub async fn new(
+        config: BigQueryConfig,
+        schema: Schema,
+        pk_indices: Vec<usize>,
+        is_append_only: bool,
+    ) -> Result<Self> {
+        let client = Client::from_service_account_key_file(&config.common.file_path)
+            .await
+            .map_err(|e| SinkError::BigQuery(e.to_string()))
+            .unwrap();
+        Ok(Self {
+            config,
+            schema: schema.clone(),
+            pk_indices,
+            client,
+            is_append_only,
+            insert_request: TableDataInsertAllRequest::new(),
+            row_encoder: JsonEncoder::new_with_big_query(
+                schema,
+                None,
+                TimestampHandlingMode::String,
+            ),
+        })
+    }
+
+    async fn append_only(&mut self, chunk: StreamChunk) -> Result<()> {
+        let mut insert_vec = vec![];
```

Comment:
From the current implementation, it seems that each `self.insert_request` only lives inside the scope of `append_only`? If so, instead of storing it as field, we may create a new local variable inside each `append_only`?

Reply:
We will use this variable in two places, either when it exceeds the maximum size or when a barrier is encountered.